packer/table_transformations/patient_diagnosis_biosource_biomaterial.py

```diff
@@ -47,77 +48,66 @@ def from_obs_df_to_pdbb_df(obs):
 
     # propagate data to lower levels and display only rows that represent the lowest level
     obs_pivot = merge_redundant_rows(obs_pivot, id_columns)
+    # update values to have correct types. it depends on previous calls, hence order dependent.
+    obs_pivot = update_datatypes(obs_pivot)
+    # fix columns order
+    obs_pivot = obs_pivot[id_columns + unq_concept_paths_ord]
+    obs_pivot = obs_pivot.rename(index=str, columns=concept_path_to_name)
     obs_pivot.reset_index(drop=True, inplace=True)
     obs_pivot = obs_pivot.rename_axis(None)
 
-    # fill NaNs with empty string
-    obs_pivot.fillna('', inplace=True)
     obs_pivot.columns.name = None
 
-    # update values to have correct types
-    obs_pivot = update_datatypes(obs_pivot)
-
-    # update date fields with format
-    # update_date_fields(obs_pivot, obs)
-
     return obs_pivot
 
 
-def drop_higher_level(lowest_level_column, data):
-    lowest_level = data[data[lowest_level_column].notnull()]
-    column_index = data.columns.get_loc(lowest_level_column)
-    for index, row in lowest_level.iterrows():
-        if column_index == 3:
-            conditions = ((data[data.columns[0]] == row[0]) &
-                          (data[data.columns[1]] == row[1]) &
-                          (data[data.columns[2]] == row[2]) &
-                          (data[data.columns[3]].isnull()))
-        elif column_index == 2:
-            conditions = ((data[data.columns[0]] == row[0]) &
-                          (data[data.columns[1]] == row[1]) &
-                          (data[data.columns[2]].isnull()))
-        elif column_index == 1:
-            conditions = ((data[data.columns[0]] == row[0]) &
-                          (data[data.columns[1]].isnull()))
-        else:
-            return
-        data.drop(data[conditions].index, inplace=True)
-
-
-def limit_rows_to_lowest_level(data, id_columns):
-    logger.info('Removing redundant rows...')
-    lowest_level_col_id = len(id_columns) - 1
-    for idx, column in enumerate(reversed(id_columns)):
-        if idx == lowest_level_col_id:
+def merge_redundant_rows(data, id_columns):
+    if data.empty:
+        return
+    # sort rows by identifying columns, merging of rows strongly depends on sorting
+    rows = data.sort_values(id_columns, na_position='last').to_dict('records')
+    result_rows = [rows[0]]
+    for row in rows[1:]:
+        row_copied = False
+        for result_row in reversed(result_rows):
+            if _is_ancestor_row(row, result_row, id_columns):
+                _copy_missing_value_to_descendant_row(row, result_row, id_columns)
+                row_copied = True
+            else:
+                break
+        if not row_copied:
+            result_rows.append(row)
+    return pd.DataFrame(result_rows)
+
+
+def _is_ancestor_row(ancestor_row_candidate, descendant_row_candidate, id_columns):
+    for id_column in id_columns:
+        if pd.isna(ancestor_row_candidate[id_column]):
```

Comment:
I have an error here, did you mean:
```suggestion
        if np.isnan(ancestor_row_candidate[id_column]):
```
?

Reply:
hm, `pd.na` sounds to me like it does broader checks including `None`.
What kind of error?

Reply:
> AttributeError: module 'pandas' has no attribute 'isna'

Reply:
https://stackoverflow.com/a/48313152

Reply:
ah, thanks!

Reply:
Addressed it here https://github.com/thehyve/transmart-packer/pull/3/commits/6f86c4c2dfd95d351e118f043f5d37446f5f4a65